v3_async_wip/v3_async_wip/sastoken.py

```diff
@@ -5,173 +5,222 @@
 
-import abc
+import asyncio
+import logging
 import time
-import urllib
-from typing import Optional, Dict
+import urllib.parse
+from typing import Dict, List, Union, Awaitable, Callable, cast
 from .signing_mechanism import SigningMechanism
 
+logger = logging.getLogger(__name__)
+
+DEFAULT_TOKEN_UPDATE_MARGIN: int = 120
+REQUIRED_SASTOKEN_FIELDS: List[str] = ["sr", "sig", "se"]
+TOKEN_FORMAT: str = "SharedAccessSignature sr={resource}&sig={signature}&se={expiry}"
+
 
 class SasTokenError(Exception):
     """Error in SasToken"""
 
     pass
 
 
-class SasToken(abc.ABC):
-    """Abstract parent class for SAS Tokens.
+class SasToken:
+    def __init__(self, sastoken_str: str) -> None:
+        """Create a SasToken object from a SAS Token string
+        :param str sastoken_str: The SAS Token string
 
-    Doesn't do much, but helps with type hints
-    """
+        :raises: ValueError if SAS Token string is invalid
+        """
+        self._token_str: str = sastoken_str
+        self._token_info: Dict[str, str] = _get_sastoken_info_from_string(sastoken_str)
 
-    @property
-    @abc.abstractmethod
-    def expiry_time(self) -> int:
-        pass
+    def __str__(self) -> str:
+        return self._token_str
 
+    @property
+    def expiry_time(self) -> float:
+        # NOTE: Time is typically expressed in float in Python, even though a
+        # SAS Token expiry time should be a whole number.
+        return float(self._token_info["se"])
 
-class RenewableSasToken(SasToken):
-    """Renewable Shared Access Signature Token used to authenticate a request.
+    @property
+    def resource_uri(self) -> str:
+        uri = self._token_info["sr"]
+        return urllib.parse.unquote(uri)
 
-    This token is 'renewable', which means that it can be updated when necessary to
-    prevent expiry, by using the .refresh() method.
+    @property
+    def signature(self) -> str:
+        signature = self._token_info["sig"]
+        return urllib.parse.unquote(signature)
 
-    Data Attributes:
-    expiry_time (int): Time that token will expire (in UTC, since epoch)
-    ttl (int): Time to live for the token, in seconds
-    """
 
-    _auth_rule_token_format = (
-        "SharedAccessSignature sr={resource}&sig={signature}&se={expiry}&skn={keyname}"
-    )
-    _simple_token_format = "SharedAccessSignature sr={resource}&sig={signature}&se={expiry}"
-
-    def __init__(
-        self,
-        uri: str,
-        signing_mechanism: SigningMechanism,
-        key_name: Optional[str] = None,
-        ttl: int = 3600,
-    ) -> None:
-        """
-        :param str uri: URI of the resource to be accessed
-        :param signing_mechanism: The signing mechanism to use in the SasToken
-        :type signing_mechanism: Child classes of :class:`azure.iot.common.SigningMechanism`
-        :param str key_name: Symmetric Key Name (optional)
-        :param int ttl: Time to live for the token, in seconds (default 3600)
+class SasTokenGenerator:
```

Comment:
Ultimately, I did end up deciding to go with an object instead of a function for this, even though it's a little more bloated. The reasoning was that I just felt the functional programming style implementation clashed with everything else being more object oriented. I think having a consistent style is worth the tradeoff of being slightly more bulky than is strictly necessary